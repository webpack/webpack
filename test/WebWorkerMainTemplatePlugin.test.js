"use strict";

const sinon = require("sinon");
const WebWorkerMainTemplatePlugin = require("../lib/webworker/WebWorkerMainTemplatePlugin");
const applyPluginWithOptions = require("./helpers/applyPluginWithOptions");

describe("WebWorkerMainTemplatePlugin", function() {
	let env;

	beforeEach(() => {
		env = {};
	});

	it("has apply function", function() {
		expect((new WebWorkerMainTemplatePlugin()).apply).toBeInstanceOf(Function);
	});

	describe("when applied", function() {
		beforeEach(function() {
			env.eventBindings = applyPluginWithOptions(WebWorkerMainTemplatePlugin);
			env.handlerContext = {
				requireFn: 'requireFn',
				indent: (value) => typeof value === 'string' ? value : value.join("\n"),
				asString: (values) => values.join("\n"),
				renderCurrentHashCode: (value) => value,
				outputOptions: {
					chunkFilename: 'chunkFilename'
				},
				applyPluginsWaterfall: (moduleName, fileName, data) => {
					return `"${moduleName}${data.hash}${data.hashWithLength()}${data.chunk && data.chunk.id || ''}"`;
				},
				renderAddModule: () => 'renderAddModuleSource();',
			};
		});

		it("binds five event handlers", function() {
			expect(env.eventBindings.length).toBe(5);
		});

		describe("local-vars handler", function() {
			beforeEach(() => {
				env.eventBinding = env.eventBindings[0];
			});

			it("binds to local-vars event", () => {
				expect(env.eventBinding.name).toBe("local-vars");
			});

			describe("when no chunks are provided", () => {
				beforeEach(() => {
					const chunk = {
						ids: [],
						chunks: []
					};
					env.source = env.eventBinding.handler.call(env.handlerContext, "moduleSource()", chunk);
				});

				it("returns the original source", () => {
					expect(env.source).toBe("moduleSource()")
				});
			});

			describe("when chunks are provided", () => {
				beforeEach(() => {
					const chunk = {
						ids: [1, 2, 3],
						chunks: [
							'foo',
							'bar',
							'baz'
						]
					};
					env.source = env.eventBinding.handler.call(env.handlerContext, "moduleSource()", chunk, 'abc123');
				});

				it("returns the original source with installed mapping", () => {
					expect(env.source).toMatchSnapshot();
				});
			});
		});

		describe("require-ensure handler", () => {
			beforeEach(() => {
				env.eventBinding = env.eventBindings[1];
			});

			it("binds to require-ensure event", () => {
				expect(env.eventBinding.name).toBe("require-ensure");
			});

			describe("when called", () => {
				beforeEach(() => {
					const chunk = {};
					env.source = env.eventBinding.handler.call(env.handlerContext, "moduleSource()", chunk, 'abc123');
				});

				it("creates import scripts call and promise resolve", () => {
					expect(env.source).toMatchSnapshot();
				});
			});
		});

		describe("bootstrap handler", () => {
			beforeEach(() => {
				env.eventBinding = env.eventBindings[2];
			});

			it("binds to bootstrap event", () => {
				expect(env.eventBinding.name).toBe("bootstrap");
			});

			describe("when no chunks are provided", () => {
				beforeEach(() => {
					const chunk = {
						ids: [],
						chunks: []
					};
					env.source = env.eventBinding.handler.call(env.handlerContext, "moduleSource()", chunk);
				});

				it("returns the original source", () => {
					expect(env.source).toBe("moduleSource()")
				});
			});

			describe("when chunks are provided", () => {
				beforeEach(() => {
					const chunk = {
						ids: [1, 2, 3],
						chunks: [
							'foo',
							'bar',
							'baz'
						]
					};
					env.source = env.eventBinding.handler.call(env.handlerContext, "moduleSource()", chunk);
				});

				it("returns the original source with chunk callback", () => {
					expect(env.source).toMatchSnapshot();
				});
			});
		});

		describe("hot-bootstrap handler", () => {
			beforeEach(() => {
				env.eventBinding = env.eventBindings[3];
			});

			it("binds to hot-bootstrap event", () => {
				expect(env.eventBinding.name).toBe("hot-bootstrap");
			});

			describe("when called", () => {
				beforeEach(() => {
					const chunk = {};
					env.source = env.eventBinding.handler.call(env.handlerContext, "moduleSource()", chunk, 'abc123');
				});

				it("returns the original source with hot update callback", () => {
					expect(env.source).toMatchSnapshot();
				});
			});
		});

		describe("hash handler", () => {
			beforeEach(() => {
				env.eventBinding = env.eventBindings[4];
				env.handlerContext = {
					outputOptions: {
						publicPath: "Alpha",
						filename: "Bravo",
						chunkFilename: "Charlie",
						chunkCallbackName: "Delta",
						library: "Echo"
					}
				};
				env.hashMock = {
					update: sinon.spy()
				};
				env.eventBinding.handler.call(env.handlerContext, env.hashMock);
			});

			it("binds to hash event", () => {
				expect(env.eventBinding.name).toBe("hash");
			});

			it("updates hash object", () => {
				expect(env.hashMock.update.callCount).toBe(7);
				sinon.assert.calledWith(env.hashMock.update, "webworker");
				sinon.assert.calledWith(env.hashMock.update, "3");
				sinon.assert.calledWith(env.hashMock.update, "Alpha");
				sinon.assert.calledWith(env.hashMock.update, "Bravo");
				sinon.assert.calledWith(env.hashMock.update, "Charlie");
				sinon.assert.calledWith(env.hashMock.update, "Delta");
				sinon.assert.calledWith(env.hashMock.update, "Echo");
			});
		});
	});
});
