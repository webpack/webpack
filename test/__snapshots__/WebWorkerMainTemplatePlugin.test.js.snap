// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebWorkerMainTemplatePlugin when applied bootstrap handler when chunks are provided returns the original source with chunk callback 1`] = `
"moduleSource()
this[\\"webpackChunk\\"] = function webpackChunkCallback(chunkIds, moreModules) {
for(var moduleId in moreModules) {
renderAddModuleSource();
}
while(chunkIds.length)
installedChunks[chunkIds.pop()] = 1;
};"
`;

exports[`WebWorkerMainTemplatePlugin when applied hot-bootstrap handler when called returns the original source with hot update callback 1`] = `
"moduleSource()
var parentHotUpdateCallback = this[\\"webpackHotUpdate\\"];
this[\\"webpackHotUpdate\\"] = function webpackHotUpdateCallback(chunkId, moreModules) {// eslint-disable-line no-unused-vars
	hotAddUpdateChunk(chunkId, moreModules);
	if (parentHotUpdateCallback) parentHotUpdateCallback(chunkId, moreModules);
} ;

function hotDownloadUpdateChunk(chunkId) {// eslint-disable-line no-unused-vars
	importScripts(requireFn.p + \\"asset-path\\" + abc123 + \\"\\" + abc123 + \\"\\" + chunkId + \\"\\");
}

function hotDownloadManifest(callback) {// eslint-disable-line no-unused-vars
	return new Promise(function (resolve, reject) {
		if (typeof XMLHttpRequest === \\"undefined\\")
		return reject(new Error(\\"No browser support\\"));
		try {
			var request = new XMLHttpRequest();
			var requestPath = requireFn.p + \\"asset-path\\" + abc123 + \\"\\" + abc123 + \\"\\";
			request.open(\\"GET\\", requestPath, true);
			request.timeout = 10000;
			request.send(null);
		} catch (err) {
			return reject(err);
		}
		request.onreadystatechange = function () {
			if (request.readyState !== 4) return;
			if (request.status === 0) {
				// timeout
				reject(new Error(\\"Manifest request to \\" + requestPath + \\" timed out.\\"));
			} else if (request.status === 404) {
				// no update available
				resolve();
			} else if (request.status !== 200 && request.status !== 304) {
				// other failure
				reject(new Error(\\"Manifest request to \\" + requestPath + \\" failed.\\"));
			} else {
				// success
				try {
					var update = JSON.parse(request.responseText);
				} catch (e) {
					reject(e);
					return;
				}
				resolve(update);
			}
		};
	});
}

function hotDisposeChunk(chunkId) {//eslint-disable-line no-unused-vars
	delete installedChunks[chunkId];
}"
`;

exports[`WebWorkerMainTemplatePlugin when applied local-vars handler when chunks are provided returns the original source with installed mapping 1`] = `
"moduleSource()

// object to store loaded chunks
// \\"1\\" means \\"already loaded\\"
var installedChunks = {
1: 1,
2: 1,
3: 1
};"
`;

exports[`WebWorkerMainTemplatePlugin when applied require-ensure handler when called creates import scripts call and promise resolve 1`] = `
"// \\"1\\" is the signal for \\"already loaded\\"
if(!installedChunks[chunkId]) {
importScripts(\\"asset-path\\" + abc123 + \\"\\" + abc123 + \\"\\" + chunkId + \\"\\");
}
return Promise.resolve();"
`;
